packages/core/src/local-workspace/s3_dir_store.ts

```diff
@@ -125,10 +126,44 @@ export const buildS3DirectoryStore = ({
       return Array.from(paths)
     }, `listing s3 objects for ${baseDir}`)
 
+  const deleteMany = async (objectPaths: string[]): Promise<void> => {
+    const objectIdentifiers: AWS.ObjectIdentifier[] = objectPaths.map(objectPath => ({ Key: getFullPath(objectPath) }))
+
+    // Split the deletion into batches of 1000 items each (AWS limit)
+    const batchSize = 1000
+    const batchPromises = []
+
+    for (let i = 0; i < objectIdentifiers.length; i += batchSize) {
+      const batch = objectIdentifiers.slice(i, i + batchSize)
+      batchPromises.push(
+        bottleneck.schedule(async () => {
+          log.trace('Deleting batch of objects from S3 bucket %s', bucketName)
+          const result = await s3.deleteObjects({
+            Bucket: bucketName,
+            Delete: { Objects: batch },
+          })
+          log.trace('Deleted batch of objects from S3 bucket %s: %s', bucketName, safeJsonStringify(result?.Deleted))
+        }),
+      )
+    }
+
+    try {
+      await Promise.all(batchPromises)
+    } catch (err) {
+      log.error('Failed to delete objects from S3 bucket %s: %s', bucketName, safeJsonStringify(err))
+      throw err
+    }
+  }
   const flush = async (): Promise<void> => {
     const files = Object.values(updated)
     updated = {}
     await Promise.all(files.map(f => writeFile(f)))
+
+    if (deleted.size > 0) {
+      const toDelete = Array.from(deleted)
+      await deleteMany(toDelete)
+      deleted = new Set()
```

Comment:
Same as for `updated`, you wouldn't like to report that you treated "delete" requests that happened during the `flush` although they weren't.
```suggestion
      deleted = new Set()
      await deleteMany(toDelete)

```